### ISSUE TITLE
IPv6 service does not works in dual-stack cluster with NodePorts enabled

### ISSUE THREAD

**Issue by Jean-Daniel** [cilium/cilium#12247]
## Bug report

I'm experimenting with a dual-stack cluster, and encounter an issue with IPv6 service reachability when cilium has `KubeProxyReplacement: NodePort` enabled.

I deployed 2 nginx services, one IPv4 and one IPv6. The IPv4 service works fine. I can `ping` it and reach it with `curl` from anywhere. For the IPv6 service, I can `ping` it normally, but `curl` never receive any reply and timeouts.

When using `cilium monitor`, I can see the initial request passing, but nothing else:

```
-> endpoint 1655 flow 0x0 identity 2->3706 state established ifindex lxcdc689d449885 orig-ip 2a01:e0a:164:2090:18b6:e0ad:fc50:5a88: [2a01:e0a:164:2090:18b6:e0ad:fc50:5a88]:52584 -> [fd00:abba:96fb::37]:80 tcp SYN
```

When disabling kube-proxy replacement, the IPv6 service works fine.

**General Information**

- Cilium version (run `cilium version`): 1.8.0
- Kernel version (run `uname -a`): 5.4.0-37-generic #41-Ubuntu 
- Orchestration system version in use (e.g. `kubectl version`, Mesos, ...): Kubernetes v1.18.4

[cilium-sysdump-20200623-235939.zip](https://github.com/cilium/cilium/files/4822199/cilium-sysdump-20200623-235939.zip)

**How to reproduce the issue**

1. instruction 1
This is a minimal cluster on bare-metal with a single controller and a single node.
This cluster only contains cilium and MetalLB. 

Note: It runs with a patched version of Kubelet that export both IPv4 and IPv6 node addresses as InternalIPs (this is how kubelet behaves on cloud providers, but bare-metal support for this feature is still a work in progress)

I installed cilium using the command:

```
helm install cilium cilium/cilium --version 1.8.0 \
--namespace kube-system \
--set global.ipv6.enabled=true \
--set global.tunnel=disabled \
--set global.autoDirectNodeRoutes=true \
--set global.nativeRoutingCIDR=10.145.0.0/16 \
--set config.ipam=cluster-pool \
--set global.ipam.operator.clusterPoolIPv4PodCIDR=10.145.0.0/16 \
--set global.ipam.operator.clusterPoolIPv6PodCIDR=fd00:abba:96fb::/104
```

The nginx service is declared as follow:

```
apiVersion: v1
kind: Service
metadata:
  name: nginx
  annotations:
    metallb.universe.tf/address-pool: default
spec:
  selector:
    app: nginx
  ports:
  - port: 80
    targetPort: 80
  type: LoadBalancer
  loadBalancerIP: 81.201.186.64
  externalTrafficPolicy: Local
---
apiVersion: v1
kind: Service
metadata:
  name: nginx6
  annotations:
    metallb.universe.tf/address-pool: ipv6
spec:
  ipFamily: IPv6
  selector:
    app: nginx
  ports:
  - port: 80
    targetPort: 80
  type: LoadBalancer
  loadBalancerIP: 2a02:2218:100:1:1eaf::10
  externalTrafficPolicy: Local
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-deployment
  labels:
    app: nginx
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:1.18.0
        ports:
        - containerPort: 80
```

**Comment by kvaps** [comment]
Hi @Jean-Daniel, I also working on dual-stack with cilium.
I was walking through [Validate IPv4/IPv6 dual-stack](https://kubernetes.io/docs/tasks/network/validate-dual-stack/) and found that nodes have no second address assigned:

```bash
# kubectl get nodes m1c35 -o go-template --template='{{range .status.addresses}}{{printf "%s: %s \n" .type .address}}{{end}}'
InternalIP: 10.28.36.175 
Hostname: m1c35 
```
@uablrek left [comment](https://github.com/kubernetes/kubernetes/issues/88784#issuecomment-595959905) in other issue that currently it is not supported by the kubelet self-registration procedure:

> The InternalIP is reported by kubelet if you specify --register-node=true. Currently kubelet does not report both ipv4 and ipv6 addresses, that is not implemented yet. But public clouds do not let kubelet report the node but defines them themselves. At least Azure set both addresses. So it depends on your environment.

**Comment by Jean-Daniel** [comment]
As I said in the original report, I'm using a patched version of kubelet that supports taking both IPv4 and IPv6 in the --node-ip parameter:

```
kubectl get nodes node-1.kube -o go-template --template='{{range .status.addresses}}{{printf "%s: %s \n" .type .address}}{{end}}'
InternalIP: 10.3.0.75 
InternalIP: 2a02:2218:100:1::1:b 
Hostname: node-1.kube 
```

And it works like a charm with kube-proxy. It fails only when trying to enable cillium optimizations.

**Comment by kvaps** [comment]
I can confirm that I'm having the same issue in similar setup

**Comment by brb** [comment]
Anyone would like to help debugging this? Basically, enabling datapath verbose msgs with the cilium-agent's `--debug-verbose=datapath` and then looking into `cilium-monitor -v` on both nodes to understand what happens with the request should give enough hints.

**Comment by kvaps** [comment]
@brb,

```
curl '[2a02:2b88:1:2:24a1:ffff:0:1]'
```

```
CPU 01: MARK 0x89737eb9 FROM 2076 DEBUG: Successfully mapped addr.p4=[::0:1] to identity=2
CPU 01: MARK 0x89737eb9 FROM 2076 DEBUG: Conntrack lookup 1/2: src=[::0:1]:41104 dst=[::0:1]:80
CPU 01: MARK 0x89737eb9 FROM 2076 DEBUG: Conntrack lookup 2/2: nexthdr=6 flags=4
CPU 01: MARK 0x89737eb9 FROM 2076 DEBUG: CT verdict: New, revnat=0
CPU 01: MARK 0x89737eb9 FROM 2076 DEBUG: Service backend slot lookup: slot=1, dport=80
CPU 01: MARK 0x89737eb9 FROM 2076 DEBUG: Conntrack create: proxy-port=0 revnat=68 src-identity=0
CPU 01: MARK 0x89737eb9 FROM 2076 DEBUG: Conntrack lookup 1/2: src=[::0:1]:41104 dst=[::0:1]:80
CPU 01: MARK 0x89737eb9 FROM 2076 DEBUG: Conntrack lookup 2/2: nexthdr=6 flags=1
CPU 01: MARK 0x89737eb9 FROM 2076 DEBUG: CT verdict: New, revnat=0
CPU 01: MARK 0x89737eb9 FROM 2076 DEBUG: Conntrack create: proxy-port=0 revnat=68 src-identity=1
CPU 01: MARK 0x89737eb9 FROM 2076 DEBUG: Conntrack create: proxy-port=0 revnat=0 src-identity=1
CPU 04: MARK 0x85916324 FROM 2076 DEBUG: Successfully mapped addr.p4=[::0:1] to identity=2
CPU 01: MARK 0x19631b4 FROM 666 DEBUG: Conntrack lookup 2/2: nexthdr=17 flags=1
CPU 01: MARK 0x19631b4 FROM 666 DEBUG: CT verdict: New, revnat=0
CPU 01: MARK 0x19631b4 FROM 666 DEBUG: Conntrack create: proxy-port=0 revnat=0 src-identity=4050 lb=0.0.0.0
CPU 03: MARK 0x88ea1873 FROM 2076 DEBUG: Conntrack lookup 2/2: nexthdr=17 flags=0
CPU 03: MARK 0x88ea1873 FROM 2076 DEBUG: CT entry found lifetime=52019, revnat=0
CPU 03: MARK 0x88ea1873 FROM 2076 DEBUG: CT verdict: Reply, revnat=0
CPU 04: MARK 0xac427298 FROM 2225 DEBUG: Conntrack lookup 2/2: nexthdr=6 flags=1
CPU 04: MARK 0xac427298 FROM 2225 DEBUG: CT entry found lifetime=52011, revnat=0
CPU 04: MARK 0xac427298 FROM 2225 DEBUG: CT verdict: Established, revnat=0
CPU 06: MARK 0x8982fd65 FROM 2225 DEBUG: Conntrack lookup 2/2: nexthdr=6 flags=1
CPU 06: MARK 0x8982fd65 FROM 2225 DEBUG: CT entry found lifetime=52011, revnat=0
CPU 06: MARK 0x8982fd65 FROM 2225 DEBUG: CT verdict: Established, revnat=0
CPU 03: MARK 0xee7d7f47 FROM 2225 DEBUG: Conntrack lookup 2/2: nexthdr=6 flags=1
CPU 03: MARK 0xee7d7f47 FROM 2225 DEBUG: CT entry found lifetime=52011, revnat=0
CPU 03: MARK 0xee7d7f47 FROM 2225 DEBUG: CT verdict: Established, revnat=0
CPU 01: MARK 0xc7303366 FROM 2225 DEBUG: Conntrack lookup 2/2: nexthdr=17 flags=1
CPU 01: MARK 0xc7303366 FROM 2225 DEBUG: CT verdict: New, revnat=0
CPU 01: MARK 0xc7303366 FROM 2225 DEBUG: Conntrack create: proxy-port=0 revnat=0 src-identity=4050 lb=0.0.0.0
CPU 01: MARK 0x1e5f7859 FROM 2076 DEBUG: Conntrack lookup 2/2: nexthdr=17 flags=0
CPU 01: MARK 0x1e5f7859 FROM 2076 DEBUG: CT entry found lifetime=52019, revnat=0
CPU 01: MARK 0x1e5f7859 FROM 2076 DEBUG: CT verdict: Reply, revnat=0
```

**Comment by brb** [comment]
Can you check whether the node which runs the backend receives the request?

**Comment by kvaps** [comment]
It is

```
12:31:35.918494 IP6 2a02:2b88:1:7:c507:c327:b8f1:1da.60126 > 2a02:2b88:1:2:24a1:ffff:0:1.80: Flags [S], seq 3620978237, win 64800, options [mss 1440,sackOK,TS val 2945995624 ecr 0,nop,wscale 7], length 0
12:31:35.918539 IP6 2a02:2b88:1:2:24a1::1.80 > 2a02:2b88:1:7:c507:c327:b8f1:1da.60126: Flags [S.], seq 782012026, ack 3620978238, win 64260, options [mss 1440,sackOK,TS val 3504437560 ecr 2945995624,nop,wscale 7], length 0
``

**Comment by kvaps** [comment]
Ah it seems problem in masquerade client sends packets to service ip `2a02:2b88:1:2:24a1:ffff:0:1` but getting response from the node ip `2a02:2b88:1:2:24a1::1`.

Standard situation "two virgins in the cold" 😀

**Comment by kvaps** [comment]
@brb, even if I remove any MASQUERADE rules node is always respond from the node ip, why so?

**Comment by kvaps** [comment]
After the reboot I see different picture:

only one packet sent, and no response packets:
```
14:39:05.783769 IP6 2a02:2b88:1:2:24af::1.41220 > 2a02:2b88:1:2:24a1:ffff:0:1.80: Flags [S], seq 3226125501, win 64800, options [mss 1440,sackOK,TS val 235245729 ecr 0,nop,wscale 7], length 0
```
`cilium monitor -v` shows the same:
```
CPU 04: MARK 0x8ea97004 FROM 870 DEBUG: Successfully mapped addr.p4=[::0:1] to identity=2
CPU 04: MARK 0x8ea97004 FROM 870 DEBUG: Conntrack lookup 1/2: src=[::0:1]:41220 dst=[::0:1]:80
CPU 04: MARK 0x8ea97004 FROM 870 DEBUG: Conntrack lookup 2/2: nexthdr=6 flags=4
CPU 04: MARK 0x8ea97004 FROM 870 DEBUG: CT verdict: New, revnat=0
CPU 04: MARK 0x8ea97004 FROM 870 DEBUG: Service backend slot lookup: slot=1, dport=80
CPU 04: MARK 0x8ea97004 FROM 870 DEBUG: Conntrack create: proxy-port=0 revnat=10 src-identity=0
CPU 04: MARK 0x8ea97004 FROM 870 DEBUG: Conntrack lookup 1/2: src=[::0:1]:41220 dst=[::0:da]:80
CPU 04: MARK 0x8ea97004 FROM 870 DEBUG: Conntrack lookup 2/2: nexthdr=6 flags=1
CPU 04: MARK 0x8ea97004 FROM 870 DEBUG: CT verdict: New, revnat=0
CPU 04: MARK 0x8ea97004 FROM 870 DEBUG: Conntrack create: proxy-port=0 revnat=10 src-identity=1
CPU 04: MARK 0x8ea97004 FROM 870 DEBUG: Conntrack create: proxy-port=0 revnat=0 src-identity=1
CPU 04: MARK 0x8ea97004 FROM 870 DEBUG: Inheriting identity=2 from stack
CPU 04: MARK 0x8ea97004 FROM 870 DEBUG: Successfully mapped addr.p4=[::0:1] to identity=2
CPU 04: MARK 0x8ea97004 FROM 870 DEBUG: Attempting local delivery for container id 2568 from seclabel 2
CPU 04: MARK 0x8ea97004 FROM 2568 DEBUG: Conntrack lookup 1/2: src=[::0:1]:41220 dst=[::0:da]:80
CPU 04: MARK 0x8ea97004 FROM 2568 DEBUG: Conntrack lookup 2/2: nexthdr=6 flags=0
CPU 04: MARK 0x8ea97004 FROM 2568 DEBUG: CT entry found lifetime=1239, revnat=0
CPU 04: MARK 0x8ea97004 FROM 2568 DEBUG: CT verdict: Established, revnat=0
```

**Comment by fristonio** [comment]
Related #13047

**Comment by fristonio** [comment]
Fixed in #13162 
I am creating a followup issue to address the fact that Dual Stack IPv6 Nodeport services don't work when running alongside kube-proxy in Iptables mode. This should not be a problem in kube-proxy free and kube-proxy with ipvs mode.
